---
title: "Single Cell - Example analysis"
author: "Miłosz Chodkowski"
output:
  html_document:
    fig_width: 6
    fig_height: 4
---

# Experiment summary

Dataset of induced pluripotent stem cells generated from three different individuals (Tung et al. 2017) in Yoav Gilad's lab at the University of Chicago. The experiments were carried out on the Fluidigm C1 platform using unique molecular identifiers (UMIs) for quantification.

GEO Accession: ***GSE77288***

# R packages

> "How hard can it be?" *\~ Jeremy Clarkson*

------------------------------------------------------------------------

## Load libraries

```{r Libs load, message=FALSE, warning=FALSE}
if(!require("SingleCellExperiment")){
  BiocManager::install("SingleCellExperiment")
  require("SingleCellExperiment")
}
if(!require("scater")){
  BiocManager::install("scater")
}
if(!require("scuttle")){
  BiocManager::install("scuttle")
}
if(!require("tidyverse")){
  install.packages("tidyverse")
  library("tidyverse")
}
if(!require("data.table")){
  install.packages("data.table")
}

```

## Load data

Original file called

```{r Original data}
df = data.table::fread("GSE77288_molecules-raw-single-per-sample.txt.gz")
df[1:5, 1:5]
```

Splitted into molecules (a tab-delimited file with the gene counts for each cell):

```{r Molecules}
molecules = read.table("data/tung/molecules.txt", sep = "\t")

molecules[1:5, 1:5]
```

And annotation for molecules (a tab-delimited text file with the cell annotations)

```{r Annotations}
annotation = read.table("data/tung/annotation.txt", sep = "\t", header = TRUE)
annotation[1:5, 1:5]
```

## Create object of type `SingleCellExperiment`

![SingleCellExperiment obj schema](https://www.singlecellcourse.org/figures/singlecellexperiment.png)

Object `SingleCellExperiment` uses a very popular `SummarizedExperiment` object (used in all biological anlyses). We pass a list to the `assays` argument in which we give our molecules data as `counts`. Then we add a `colData` parameter to describe our *samples*. We can optionally add `rowData` which is a parameter to describe our *genes.*

```{r SCE init}

experiment = SingleCellExperiment(
  assays = list(counts = as.matrix(molecules)), # Type required is matrix, list passed as kwargs to `SummarizedExperiment)
  colData = annotation # Can be table
)
experiment
```

According to schema, data describing our samples are in colData. Let's see it:

```{r Samples description}
colData(experiment)
```

We didn't specify any additional information for our genes, so rowData should be an empty df with named rows only:

```{r Genes description}
print(rowData(experiment))
print(rowData(experiment) %>% rownames() %>% head(5)) # First 5 genes
```

## Some stats

```{r Log assay}
# Add a new assay that was log transformed
# We can use `assay(sce, "assay_name")<-` to get or set a new assay
# Or we can use some predefined functions with predefined names
# Examples: `counts` - raw counts, `normcounts` - normalized counts, `logcounts` - logtransformed counts, `tpm` - transcripts per million, etc.

# We already have raw counts
# counts(experiment)

# So just set new logcounts3
logcounts(experiment) <- log2(counts(experiment) + 1) # Add 1 to prevent log(0) = -Inf
logcounts(experiment)[1:5, 1:5]
# We can also use another method from another package that just works on SingleCellExperiment obj
experiment = scuttle::logNormCounts(experiment)
```

We can also add some mean count and total count per cell

```{r Add mean counts}
print(counts(experiment)[1:5, 1:5])
cat("\n\n")
print(colMeans(counts(experiment)[1:5, 1:5]))
cat("\n\n")
colData(experiment)$mean_counts = colMeans(counts(experiment))
colData(experiment)$total_counts = colSums(counts(experiment))
cat("\n\n")
print(colData(experiment)[1:5, ])
```

## Visual presentation

This is utterly f\*\*\*\*\*\* disgusting *\~ Python programmer when he saw the R plotting code*

------------------------------------------------------------------------

Let's plot some stats for each batch:

```{r Some stats for each bach}
cells = colData(experiment) %>% as.data.frame() # convert from c(Data.frame, SingleCellExperiment) to data.frame
print(cells[1:5, 1:5])

th = theme(axis.text.x  = element_text(angle = -45, hjust = -0.1))

p = ggplot(data = cells)
p + geom_violin(aes(x = batch, mean_counts, fill = batch)) + th
p + geom_violin(aes(x = batch, total_counts, fill = batch)) + th

```

If we we want to plot some data from specific assay showing a specific *GENE*, then we can use `ggcells` function from `scater` package:

```{r Specific assay plot}
# Using scater's ggcells we can give a SingleCellExperiment instead of data.frame
# Above example using scater
scater::ggcells(experiment, aes(x = batch, y = mean_counts, fill = batch)) + geom_violin() + th
# individual gene
# expr_values default = logcounts
scater::ggcells(experiment, aes(x = batch, y = ENSG00000198938, fill = batch), exprs_values = "logcounts") + geom_violin() + th

# And scatter plot
# First calculate in each sample
colData(experiment)$var_counts <- colVars(counts(experiment))
colData(experiment)

# Now plot
scater::ggcells(experiment, aes(x = mean_counts, y = var_counts, color = batch)) + 
  geom_point() + 
  stat_smooth(method = "lm", aes(x = mean_counts, y = var_counts, color = NULL)) +
  ggpubr::stat_cor(aes(label = after_stat(rr.label), color = NULL))
              
```

##### Ok chief but what about interpretation?

> Positive correlation between mean(counts) and var(counts) is typical for RNA-seq, especially raw count data. This is not a good scenario (it's not really problematic if we know how to deal with it), because a lot of expression/FC estimators and test-s producing p-values and q-values assume that mean and variance are independent variables (no correlation).

### Sensational dimensionality reduction

`Scater` library and `SingleCellExperiment` are tightly bound together, because `Scater` is based on `SingleCellExperiment`. It is very easy to apply some dimensionality reduction techniques and plot some advanced visualization *JUST LIKE THAT*. Let's try to do some dimensionality reduction:

```{r Dim reduction}
experiment = scater::runPCA(experiment, ncomponents = 2) # Returns new experiment with a new reducedDim object
print(experiment) # reducedDimNames(1) -> Now containes new entry called "PCA"
#  Access reduced dimname using reducedDim() functin
reducedDim(experiment, "PCA")[1:10, ] # first ten rows for clarity
# Maybe plot some PCA?
# First create some plots 
# Colour_by is a name of the column in colData -> Remember? colData is data.frame with additional information for each cell. 
# We've performed PCA to reduce dimensionality for each *CELL*. 
# So no we plot each cell and we can add some information -> that's why I used colour_by = "batch"

batch_pca = scater::plotReducedDim(experiment, "PCA", colour_by = "batch")
ind_pca = scater::plotReducedDim(experiment, "PCA", colour_by = "individual")

ggpubr::ggarrange(batch_pca, ind_pca, labels = c("PCA colored by batch", "PCA colored by individual"), font.label = list(size = 12))

```

So we got the PCA, but it's not widely used in single-cell. Let's do umap and tSNE colored by individual

```{r UMAPs and TSNEs}
experiment = experiment %>% 
  scater::runUMAP() %>%
  scater::runTSNE()
experiment %>% print() # reducedDimNames(3) -> got new entries, named "UMAP" and "TNSE" 

# Lets see new reduced dim
reducedDim(experiment, "UMAP") %>% head() %>% print
reducedDim(experiment, "TSNE") %>% head() %>% print()
# Ok so let's plot it
ind_umap = scater::plotReducedDim(experiment, "UMAP", colour_by = "individual")
ind_tsne = scater::plotReducedDim(experiment, "TSNE", colour_by = "individual")
ggpubr::ggarrange(ind_umap, ind_tsne)

```

Even more reduced dimensions

```{r NMF and MDS}
# Ok let's do more (lok it's so easy idk why people hate R)
experiment = experiment %>% 
  scater::runNMF() %>%
  scater::runMDS()
experiment %>% print() # Lol 2 new entries
# so now we have all 5 reduced dim entries
reducedDimNames(experiment) %>% print() # [1] "PCA"  "UMAP" "TSNE" "NMF"  "MDS" 

# plot last 2 of them
ind_nmf = scater::plotReducedDim(experiment, "NMF", colour_by = "individual")
ind_mds = scater::plotReducedDim(experiment, "MDS", colour_by = "individual")
ggpubr::ggarrange(ind_nmf, ind_mds)

```

# Real analysis (Dataset load + QC + Low expression filtering)

> Just do it! \*\~ ur transcriptomics teacher\*

---

## Dataset
```{r Libs, message=FALSE, warning=FALSE}
# Scater with AnnotationDbi and org.Hs.ed.db to convert EnsemblIDs to gene names
library(scater)
library(SingleCellExperiment)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(EnsDb.Hsapiens.v86)


# In bioconductor we can have a lot of annotation resources. The are Database packages (like normal databases)
# These resources can be GeneCentric or GenomeCentric (annotation on individual genes or whole genomes respectively)
# Gene centric annotations can be:
#   - on organism level: org.Mm.eg.db
#   - on platform level: hgu133plus2.db
#   - on system   level: GO.db
# Genome centric annotations can be:
#   - on transcriptome level:  TxDb.Hsapiens.UCSC.hg19.knownGene (Tx -transcript, Db- database = TranscriptDatabase)
#   - on generic genome features: via GenomicFeatures package
# Resources can also be web-based:
#   - biomart: via biomaRt package

# AnnotationDbi ("i" - interface) is a package that can easily query those databases
# So we have Scater for some useful data analysis, db packages for getting metadata on our genes/genomes
# and AnnotationDbi for all the queries

# AnnotationDbi has methods like normal SQL language (SELECT, etc)
# DB packages, after loading expose an object named exactly the same as the package
# We can use the methods from AnnotationDbi to query db object using SQL methods
# 
# Example:
# Types of columns our data has
columns(org.Hs.eg.db) %>% print()
# Types of keys that we can use to fetch the data
keytypes(org.Hs.eg.db) %>% print()
# Fetch keys using "SYMBOL" keytype and print some first keys
keys(org.Hs.eg.db, keytype = "SYMBOL") %>% head() %>% print()
# Or another using "ENSEMBL"
keys(org.Hs.eg.db, keytype = "ENSEMBL") %>% head() %>% print()
# Lets say we want to fetch first 15 genes
# Lets get the first 15 keys (let's use symbol)
qkeys = keys(org.Hs.eg.db, keytype = "SYMBOL") %>% head(15)
print(qkeys)
# SELECT on org.Hs.eg.db using our keys, keytype and return a columns
fetched = select(org.Hs.eg.db, keys = qkeys, keytype = "SYMBOL", columns = c("SYMBOL", "ENSEMBL"))
print(fetched)

```

Create our experiment once again. Let's name it ***sce***
```{r New SCE}
sce = SingleCellExperiment(assays = list(counts = as.matrix(molecules)), colData = annotation)
print(sce)
```

Now we have to do some magic -> remove RNA spike-ins. But what are they? Acutally RNA spike-ins are synthetic RNA molecules that are added to a sample at a known concentration in order to control for variability in RNA sequencing experiments. They can be used to assess the accuracy and precision of the RNA-sequencing method, and to normalize the data from different samples.

RNA spike-ins are typically made of RNA sequences that are well characterized and have known properties, such as length, GC content, and secondary structure. They can be derived from a variety of sources, including synthetic RNA oligonucleotides or cDNAs, and are typically added to the sample at a concentration that is much lower than the endogenous RNA present in the sample.

So, we have to remove spike-ins from regular assays and store them as an 'alternative experiment'. Typical examples besided spike-in transcripts can be CRISPR tags in CITE-seq experiments. These data cannot be stored in the main assays of the SingleCellExperiment itself.

```{r RM ERCC}
# In our dataset spike-ins are annotated as ERCC 
# (ERCC RNA spike-in mix, Invitrogen™, ThermoFisher Scientific)
# So (according to SingleCellExperiment schema) we should remove genes (rows)
# that begins with ERCC

# Let's see if we have any
# Get gene names (from rows)
rownames(sce) %>%
  # Construct lambda expr to check if gene name starts with ERCC
  purrr::keep(~ stringr::str_starts(.x, "ERCC")) %>% 
  # Take some first
  head() %>%
  # Print it
  print()
# Okay so we have some. Now let's extract them from regular assay to alternativeExperiment slot
# Remember! we can index SingleCellExperiment obj using gene names and sample names
ercc_idx = grep(pattern = "^ERCC", rownames(sce)) # indicies of ERCC transcripts
print(ercc_idx %>% length()) # 89 ERCC's, nice 

altExp(sce, "ERCC") <- sce[ercc_idx, ] # Set new alternativeExperiment by indexing sce
print(sce) # altExpNames(1): ERCC - nice

# Now remove these ERCC's from regular assays
sce = sce[-ercc_idx, ] # rownames(18938), previously (19027) - nice

```

Now, let’s map ENSEMBL IDs to gene symbols. 
```{r Mam Ensembls}
# 'mapIds' gets the mapped ids (column) for a set of keys that are of a particular keytype. Usually returned as a named character vector.
# In other words this function queries that db with our keys and returns a column that is mapped to our keys
# So we get something like a dictionary, where key: our serach key, value: returned values
# In this example we use ENSEMBL_IDS from our dataset as keys, and returned value is vector os gene symbols
# where names are our keys. It's like a map, u know XD
gene_names <- mapIds(org.Hs.eg.db, keys=rownames(sce), keytype="ENSEMBL", column="SYMBOL")
print(gene_names %>% head())
rowData(sce)$SYMBOL <- gene_names # Add new rowData entry
print(rowData(sce))
# Check for na 
table(is.na(gene_names))
```
From the table command, we can see that most genes were annotated; however, 860 returned “NA.” By default, mapIds returs one symbol per ID; this behaviour can be changed using multiVals argument( we can take first, last, pass some function, etc). Let's remove all those NA's.
```{r idk}
sce = sce[!is.na(rowData(sce)$SYMBOL), ]
print(sce)
```

## Basic QC

> The following scater functions allow us to add per-cell and per-gene metrics useful for dataset evaluation. Most popular metrics per cell are total number of counts (UMIs), total number of detected genes, total number of mitochondrial counts, percent of mitochondrial counts, etc.

```{r Compute basic metrics}
umi_cell <- perCellQCMetrics(sce, use.altexps = T) # Also perform QC for alternative experiments

# sum - sum of counts per cell
# detected - number of detected genes
# percent - percentage of counts in the top features
print(umi_cell)

umi_feature <- perFeatureQCMetrics(sce)
# mean - mean counts per gene
# deteced - number of gene detections
print(umi_feature)

# Add these QC metrics to our SingleCellExperiment object
sce = sce %>% 
  addPerCellQC(use.altexps = T) %>%
  addPerFeatureQC()
print(sce) # We Have more colData(), and rowData()
print(colData(sce)[1:10, ]) 
# Total column means total number of detected genes in a sample


print(rowData(sce)[1:10, ])

```
Now plots some QC stats.
```{r Plots basic stats on QC}
# Plot a histogram of total number of counts per sample
# @ is slot, or quicker access to property
total_hist = ggplot(data = sce@colData %>% as.data.frame(), aes(total)) + geom_histogram(color = "red")
p1 = total_hist + ggtitle("Histogram of total number of counts")

# Plot hist of total number of detected genes per sample
total_detected = ggplot(data = sce@colData %>% as.data.frame(), aes(detected)) + geom_histogram(color = "green")
p2 = total_detected + ggtitle("Histogram of total number of detected genes")

ggpubr::ggarrange(p1, p2)
```
## Filter low "expressed" cells

Sometimes it’s hard to come up with an obvious filtering cutoff. In this case, adaptive threshold can help us identify points that are more than 3 median absolute deviations (MADs) away from the median in any of the variables we use for QC

```{r Compute outliers}
# Exclude cells where sum of counts if too loew
th_sum = isOutlier(sce@colData$sum, log = TRUE, type = "lower") 
# Exclude cells where total number of detected genes is too low
th_detected = isOutlier(sce@colData$detected, log = TRUE, type = "lower")
# Exclude cells where ERCC percent is too high
th_ercc = isOutlier(sce@colData$altexps_ERCC_percent, log = TRUE, type = "higher") 

discarded = th_sum | th_detected | th_ercc # Union of all the logical vectors to create one to rule them all

# Let's see how many we have
DataFrame(
  LibSize=sum(th_sum), 
  NExprs=sum(th_detected), 
  SpikeProp=sum(th_ercc), 
  Total=sum(discarded)
) %>% print()
```
All of above could be done in one step XDDDD
```{r Explain shorter version of QC}
reasons <- quickPerCellQC(
  umi_cell, 
  sum.field = "sum", 
  detected.field = "detected",
  sub.fields=c("altexps_ERCC_percent"))
print(reasons)
colSums(as.matrix(reasons)) %>% print()
#              low_lib_size            low_n_features    high_altexps_ERCC_percent                   discard 
#                        47                       65                           137                       165
```
So according to these QC metrics we have to discard 165 cells. But let's store these information in our colData for now, and don't remove it just yet.
```{r Add discarded info}
sce$discard <- reasons$discard

# Just plot which cells to discard
# Use scater
s_d = scater::plotColData(sce, x = "sum", y = "detected", colour_by = "discard")
s_perc = scater::plotColData(sce, x = "sum", y = "altexps_ERCC_percent", colour_by = "discard")
ggpubr::ggarrange(s_d, s_perc, 
                  labels = c("Sum-Detected discards", "Sum-ERCC_percent discards"), 
                  common.legend = T,
                  widths = c(1.5, 2))
```

## Highly expressed genes
Let’s take a look at the most expressed genes in the whole dataset. We will use symbols we obtained using AnnotationDbi.
```{r Plot highly expressed genes}
# Use sce, counts assay as the matrix used to get the top expressed genes
# n: how many top genes?
# feature_names_to_plot: NULL - use rownames (it means we use original loaded gene names). 
# In this case we passed another column from rowData() specifying gene SYMBOL, fetched using AnnotationDbi.
# colour_cells_by: (colour CELLS, which mean we pass a metadata of cells -> column of colData()) 
p = plotHighestExprs(sce,
                 n = 25,
                 exprs_values = "counts", 
                 feature_names_to_plot = "SYMBOL", 
                 colour_cells_by="detected")
p + ggtitle("Top 25 expressed genes across all cells")
```

We see a lot of genes like "RP*..blah" which are ribosomal genes. It's normal for sc-dataset to containg a lot of ribosomal or mitochondrial genes. Let’s keep the genes which were only detected (expression value > 1) in 2 or more cells. We’ll discard approximately 4,000 weakly expressed genes thanks to this.
```{r Filter out low expressed genes}
# By row to use genes not cells
# Detection_limit = 1 -> Use only genes that have >= 1 read aligned (1 count)

# nexpr function returns number of cells in which genes after above filtering are present
# and we want only genes that are expressed in 2 cells or more -> that's why >=2 is present
keep_feature <- nexprs(sce, byrow = TRUE, detection_limit = 1) >= 2
rowData(sce)$discard <- !keep_feature # Add a 'discard' annotation to each gene (rowData())
print(table(rowData(sce)$discard))
```
Wow - we have succesfully annotated all cell and marked 4205 as lowly expressed!

# Sensational dimensionality reduction (again)
Let's create a log2 raw transformed assay now and create a SingleCellExperiment object that will apply previous QC.
```{r Add raw logcounts}
assay(sce, "logcounts_raw") <- log2(counts(sce) + 1)
print(sce) # assays(2): counts logcounts

sce.qc = sce[!rowData(sce)$discard, !colData(sce)$discard]
sce.qc
```
With log-transformation, we equalize the large difference between strongly and weakly expressed genes, and immediately see cells form groups by replicate, individual, and sequencing depth. Proof:

##### PCA on raw counts
```{r PCA on raw counts}

# Again - calculate reduced dim
sce.tmp = scater::runPCA(sce.qc, exprs_values = "counts")
plotPCA(sce.tmp, colour_by = "batch", shape_by = "individual")
```

#####PCA on log-transformed counts
```{r PCA on raw logcounts}

# Again - calculate reduced dim
sce.qc = scater::runPCA(sce.qc, exprs_values = "logcounts_raw")
plotPCA(sce.qc, colour_by = "batch", shape_by = "individual")

```
Clearly log-transformation is benefitial for our data - it reduces the variance on the first principal component and already separates some biological effects. Moreover, it makes the distribution of the expression values more normal.
But raw logcounts are not good enough. We should use the logcounts slot (produced by scater) which additionally to log2 normalizes by a lib size. Let's do it.
```{r Produce logNormCounts}
# let's use sce for clarity
# from now on `sce` object is the same as `sce.qc` (it has applied quality control)
sce = scater::logNormCounts(sce.qc)
sce # assays(3) counts logcounts_raw logcounts
```

Let's run PCA again, this time on logNormCounts
```{r PCA on logNormCounts}
sce = runPCA(sce, exprs_values = "logcounts")
plotPCA(sce, colour_by = "batch", shape_by = "individual")

```
***NOW WE'RE TALKIN'***
Let's do UMAP and tSNE
```{r UMAPs and TSNEs on logNormCounts}

sce = sce %>%
  runTSNE(exprs_values = "logcounts") %>%
  runUMAP(exprs_values = "logcounts")
u = plotUMAP(sce, colour_by = "batch", shape_by = "individual")
t = plotTSNE(sce, colour_by = "batch", shape_by = "individual")
ggpubr::ggarrange(u, t)
```
## Other steps (skipped)
We could do a couple of other steps also like:
1. Analyze correlations with PCs
2. Analyze and plot some explenatory variables
3. blah...blah...blah

> In other words we can search for "Confounding Factors". For the *sake of teacher, other students and my mental health* I'll skip that in this project. Let's do something more fun.

# Biological analysis
## Clustering
```{r Clustering using sc3}
# Even more libs
library(SC3)

# sc3() requires rowData() to have a column named 'feature_symbol' so create it
rowData(sce)$feature_symbol = rowData(sce)$SYMBOL # just copy existing
# Biology =T, kse = 3 (biology: dont add additional metadata to use with other plotting functions, kse: number of clusters)
sce.clus = sc3(sce, ks = 3, biology = T, n_cores = 4)
sce.clus
```

## Plot PCA with cluster annotation
```{r PCA with clusters}

plotPCA(sce.clus, colour_by = "sc3_3_clusters")

```


## Plot consensus map
> It represents similarity between the cells based on the averaging of clustering results from all combinations of clustering parameters. Similarity 0 (blue) means that the two cells are always assigned to different clusters. In contrast, similarity 1 (red) means that the two cells are always assigned to the same cluster.

```{r Consensus map}
sc3_plot_consensus(
    sce.clus, k = 3,
    show_pdata = c(
        "individual"
    )
)
```

## Plot DE genes
```{r Plot DE genes}
# https://github.com/hemberg-lab/SC3/issues/84
# When having rownames as ensembl ids (non unique) there are errors
# idk why
# but
# lets convert

rownames(sce.clus) = rowData(sce.clus)$feature_symbol

sc3_plot_de_genes(
  sce.clus, 
  k = 3,
  show_pdata = c(
        "individual"
    ))
```

## Plot markers

>To find marker genes, for each gene a binary classifier is constructed based on the mean cluster expression values. The classifier prediction is then calculated using the gene expression ranks. The area under the receiver operating characteristic (ROC) curve is used to quantify the accuracy of the prediction. A p-value is assigned to each gene by using the Wilcoxon signed rank test. By default the genes with the area under the ROC curve (AUROC) > 0.85 and with the p-value < 0.01 are selected and the top 10 marker genes of each cluster are visualized in this heatmap.

```{r}
sc3_plot_markers(sce.clus, k = 3)
```

# END
