---
title: "Single Cell - Example analysis"
author: "MiÅ‚osz Chodkowski"
output:
  html_document:
    fig_width: 16
    fig_height: 9
---

# Experiment summary

Dataset of induced pluripotent stem cells generated from three different individuals (Tung et al. 2017) in Yoav Gilad's lab at the University of Chicago. The experiments were carried out on the Fluidigm C1 platform using unique molecular identifiers (UMIs) for quantification.

GEO Accession: ***GSE77288***

# Load libraries

```{r}
if(!require("SingleCellExperiment")){
  BiocManager::install("SingleCellExperiment")
  require("SingleCellExperiment")
}
if(!require("scater")){
  BiocManager::install("scater")
  require("scater")
}
if(!require("tidyverse")){
  install.packages("tidyverse")
  library("tidyverse")
}
if(!require("data.table")){
  install.packages("data.table")
}

```

# Load data

Original file called

```{r}
df = data.table::fread("GSE77288_molecules-raw-single-per-sample.txt.gz")
df[1:5, 1:5]
```

Splitted into molecules (a tab-delimited file with the gene counts for each cell):

```{r}
molecules = read.table("data/tung/molecules.txt", sep = "\t")

molecules[1:5, 1:5]
```

And annotation for molecules (a tab-delimited text file with the cell annotations)

```{r}
annotation = read.table("data/tung/annotation.txt", sep = "\t", header = TRUE)
annotation[1:5, 1:5]
```

# Create object of type `SingleCellExperiment`

![SingleCellExperiment obj schema](https://www.singlecellcourse.org/figures/singlecellexperiment.png)

Object `SingleCellExperiment` uses a very popular `SummarizedExperiment` object (used in all biological anlyses). We pass a list to the `assays` argument in which we give our molecules data as `counts`. Then we add a `colData` parameter to describe our *samples*. We can optionally add `rowData` which is a parameter to describe our *genes.*

```{r}

experiment = SingleCellExperiment(
  assays = list(counts = as.matrix(molecules)), # Type required is matrix, list passed as kwargs to `SummarizedExperiment)
  colData = annotation # Can be table
)
experiment
```

According to schema, data describing our samples are in colData. Let's see it:

```{r}
colData(experiment)
```

We didn't specify any additional information for our genes, so rowData should be an empty df with named rows only:

```{r}
print(rowData(experiment))
print(rowData(experiment) %>% rownames() %>% head(5)) # First 5 genes
```

# Some stats

```{r}
# Add a new assay that was log transformed
# We can use `assay(sce, "assay_name")<-` to get or set a new assay
# Or we can use some predefined functions with predefined names
# Examples: `counts` - raw counts, `normcounts` - normalized counts, `logcounts` - logtransformed counts, `tpm` - transcripts per million, etc.

# We already have raw counts
# counts(experiment)

# So just set new logcounts
logcounts(experiment) <- log2(counts(experiment) + 1) # Add 1 to prevent log(0) = -Inf
logcounts(experiment)[1:5, 1:5]
```

We can also add some mean count and total count per cell

```{r}
print(counts(experiment)[1:5, 1:5])
cat("\n\n")
print(colMeans(counts(experiment)[1:5, 1:5]))
cat("\n\n")
colData(experiment)$mean_counts = colMeans(counts(experiment))
colData(experiment)$total_counts = colSums(counts(experiment))
cat("\n\n")
print(colData(experiment)[1:5, ])
```

# Visual presentation

Let's plot some stats for each batch:

```{r}
cells = colData(experiment) %>% as.data.frame() # convert from c(Data.frame, SingleCellExperiment) to data.frame
print(cells[1:5, 1:5])

th = theme(axis.text.x  = element_text(angle = -45, hjust = -0.1))

p = ggplot(data = cells)
p + geom_violin(aes(x = batch, mean_counts, fill = batch)) + th
p + geom_violin(aes(x = batch, total_counts, fill = batch)) + th

```

If we we want to plot some data from specific assay showing a specific *GENE*, then we can use `ggcells` function from `scater` package:

```{r}
# Using scater's ggcells we can give a SingleCellExperiment instead of data.frame
# Above example using scater
scater::ggcells(experiment, aes(x = batch, y = mean_counts, fill = batch)) + geom_violin() + th
# individual gene
# expr_values default = logcounts
scater::ggcells(experiment, aes(x = batch, y = ENSG00000198938, fill = batch), exprs_values = "logcounts") + geom_violin() + th

# And scatter plot
# First calculate in each sample
colData(experiment)$var_counts <- colVars(counts(experiment))
colData(experiment)

# Now plot
scater::ggcells(experiment, aes(x = mean_counts, y = var_counts, color = batch)) + 
  geom_point() + 
  stat_smooth(method = "lm", aes(x = mean_counts, y = var_counts, color = NULL)) +
  ggpubr::stat_cor(aes(label = after_stat(rr.label), color = NULL))
              
```

### Means what?

Positive correlation between mean(counts) and var(counts) is typical for RNA-seq, especially raw count data. This is not a good scenario (it's not really problematic if we know how to deal with it), because a lot of expression/FC estimators and test-s producing p-values and q-values assume that mean and variance are independent variables (no correlation).

# Reduced Dims

`Scater` library and `SingleCellExperiment` are tightly bound together, because `Scater` is based on `SingleCellExperiment`. It is very easy to apply some dimensionality reduction techniques and plot some advanced visualization *JUST LIKE THAT*. Let's try to do some dimensionality reduction:

```{r}
experiment = scater::runPCA(experiment, ncomponents = 2) # Returns new experiment with a new reducedDim object
print(experiment) # reducedDimNames(1) -> Now containes new entry called "PCA"
#  Access reduced dimname using reducedDim() functin
reducedDim(experiment, "PCA")[1:10, ] # first ten rows for clarity
# Maybe plot some PCA?
# First create some plots 
# Colour_by is a name of the column in colData -> Remember? colData is data.frame with additional information for each cell. 
# We've performed PCA to reduce dimensionality for each *CELL*. 
# So no we plot each cell and we can add some information -> that's why I used colour_by = "batch"

batch_pca = scater::plotReducedDim(experiment, "PCA", colour_by = "batch")
ind_pca = scater::plotReducedDim(experiment, "PCA", colour_by = "individual")

ggpubr::ggarrange(batch_pca, ind_pca, labels = c("PCA colored by batch", "PCA colored by individual"), font.label = list(size = 12))

```

So we got the PCA, but it's not widely used in single-cell. Let's do umap and tSNE colored by individual

```{r}
experiment = experiment %>% 
  scater::runUMAP() %>%
  scater::runTSNE()
experiment %>% print() # reducedDimNames(3) -> got new entries, named "UMAP" and "TNSE" 

# Lets see new reduced dim
reducedDim(experiment, "UMAP") %>% head() %>% print
reducedDim(experiment, "TSNE") %>% head() %>% print()
# Ok so let's plot it
ind_umap = scater::plotReducedDim(experiment, "UMAP", colour_by = "individual")
ind_tsne = scater::plotReducedDim(experiment, "TSNE", colour_by = "individual")
ggpubr::ggarrange(ind_umap, ind_tsne)

```

More XD

```{r}
# Ok let's do more (lok it's so easy idk why people hate R)
experiment = experiment %>% 
  scater::runNMF() %>%
  scater::runMDS()
experiment %>% print() # Lol 2 new entries - sensational
# so now we have all 5 reduced dim entries
reducedDimNames(experiment) %>% print() # [1] "PCA"  "UMAP" "TSNE" "NMF"  "MDS" 

# plot last 2 of them
ind_nmf = scater::plotReducedDim(experiment, "NMF", colour_by = "individual")
ind_mds = scater::plotReducedDim(experiment, "MDS", colour_by = "individual")
ggpubr::ggarrange(ind_nmf, ind_mds)

```
